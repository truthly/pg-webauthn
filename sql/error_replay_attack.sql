--
-- In this test we will simulate different "replay attacks",
-- by reusing values that are guaranteed to be unique
-- thanks to primary keys and unique constraints on columns.
--
-- We will use the SAVEPOINT ... ROLLBACK TO PostgreSQL feature,
-- to let errors happen that we want to test for,
-- but to roll back to the state before the error,
-- allowing the rest of the test to continue
-- which depends on the data produced so far.
-- Otherwise we would have to test all steps from start
-- in separate test files, which would be unnecessarily tedious.
--
-- See: https://www.postgresql.org/docs/current/sql-rollback-to.html
--

BEGIN;

CREATE EXTENSION IF NOT EXISTS webauthn CASCADE;

SELECT jsonb_pretty(webauthn.init_credential(
  challenge := '\x30d5843dc347fe7d9328675e964b7efded1f3112bb0f928e6fb147fc86c564ba'::bytea,
  user_name := 'alex.p.mueller@example.com',
  user_id := '\x927f90c2323748d1e2c24c39af7d240423f5d1d0fbed1d77ccfe372b2d67ab244f988bf5cf7c72f8b7f0fdcb359498a935c26f4b8fc1e2b564b54181a83468da'::bytea,
  user_display_name := 'Alex P. Müller',
  relying_party_name := 'ACME Corporation',
  relying_party_id := NULL,
  user_verification := 'discouraged',
  timeout := '00:05:00'::interval,
  challenge_at := '2020-12-15 08:08:56.596311+01'
));

SAVEPOINT init_credential;
SELECT jsonb_pretty(webauthn.init_credential(
  challenge := '\x30d5843dc347fe7d9328675e964b7efded1f3112bb0f928e6fb147fc86c564ba'::bytea,
  user_name := 'alex.p.mueller@example.com',
  user_id := '\x927f90c2323748d1e2c24c39af7d240423f5d1d0fbed1d77ccfe372b2d67ab244f988bf5cf7c72f8b7f0fdcb359498a935c26f4b8fc1e2b564b54181a83468da'::bytea,
  user_display_name := 'Alex P. Müller',
  relying_party_name := 'ACME Corporation',
  relying_party_id := NULL,
  user_verification := 'discouraged',
  timeout := '00:05:00'::interval,
  challenge_at := '2020-12-15 08:08:56.596311+01'
));
ROLLBACK TO init_credential;

SELECT * FROM webauthn.make_credential(
  credential_id := 'AXNBRMEOFaYGaROrEph1sOZ4kftILi9ry8vCw2fPQf712glIpQDRX-7HBQ2VmQVpRWU3A6Cu_XcKbnoC2SSy5_o0Z2qO7Owdnms8K0GsiqvWx3WtUPn0a8Ga6QWbkEvsUXOp9ikZ9v4DeYeTzzp0h2uAlx8ezayuqjB_uMQyB5kBVwRhkhZEmzQCl097',
  credential_type := 'public-key',
  attestation_object := 'o2NmbXRkbm9uZWdhdHRTdG10oGhhdXRoRGF0YVkBEUmWDeWIDoxodDQXD2R2YFuP5K65ooYyx5lc87qDHZdjRV_YYQytzgACNbzGCmSLCyXx8FUDAI0Bc0FEwQ4VpgZpE6sSmHWw5niR-0guL2vLy8LDZ89B_vXaCUilANFf7scFDZWZBWlFZTcDoK79dwpuegLZJLLn-jRnao7s7B2eazwrQayKq9bHda1Q-fRrwZrpBZuQS-xRc6n2KRn2_gN5h5PPOnSHa4CXHx7NrK6qMH-4xDIHmQFXBGGSFkSbNAKXT3ulAQIDJiABIVggxMJIqQI-xxUSsJGR9HIbIZHKIiN-alA9B2SdzIzjzwkiWCCHOhTTOTy3-H2rNbbfs8jojm8AeaNYEfQCplT5aYjwtA',
  client_data_json := 'eyJ0eXBlIjoid2ViYXV0aG4uY3JlYXRlIiwiY2hhbGxlbmdlIjoiTU5XRVBjTkhfbjJUS0dkZWxrdC1fZTBmTVJLN0Q1S09iN0ZIX0liRlpMbyIsIm9yaWdpbiI6Imh0dHA6Ly9sb2NhbGhvc3QiLCJjcm9zc09yaWdpbiI6ZmFsc2UsIm90aGVyX2tleXNfY2FuX2JlX2FkZGVkX2hlcmUiOiJkbyBub3QgY29tcGFyZSBjbGllbnREYXRhSlNPTiBhZ2FpbnN0IGEgdGVtcGxhdGUuIFNlZSBodHRwczovL2dvby5nbC95YWJQZXgifQ',
  credential_at := '2020-12-15 08:09:00.20485+01'
);

SAVEPOINT make_credential;
SELECT * FROM webauthn.make_credential(
  credential_id := 'AXNBRMEOFaYGaROrEph1sOZ4kftILi9ry8vCw2fPQf712glIpQDRX-7HBQ2VmQVpRWU3A6Cu_XcKbnoC2SSy5_o0Z2qO7Owdnms8K0GsiqvWx3WtUPn0a8Ga6QWbkEvsUXOp9ikZ9v4DeYeTzzp0h2uAlx8ezayuqjB_uMQyB5kBVwRhkhZEmzQCl097',
  credential_type := 'public-key',
  attestation_object := 'o2NmbXRkbm9uZWdhdHRTdG10oGhhdXRoRGF0YVkBEUmWDeWIDoxodDQXD2R2YFuP5K65ooYyx5lc87qDHZdjRV_YYQytzgACNbzGCmSLCyXx8FUDAI0Bc0FEwQ4VpgZpE6sSmHWw5niR-0guL2vLy8LDZ89B_vXaCUilANFf7scFDZWZBWlFZTcDoK79dwpuegLZJLLn-jRnao7s7B2eazwrQayKq9bHda1Q-fRrwZrpBZuQS-xRc6n2KRn2_gN5h5PPOnSHa4CXHx7NrK6qMH-4xDIHmQFXBGGSFkSbNAKXT3ulAQIDJiABIVggxMJIqQI-xxUSsJGR9HIbIZHKIiN-alA9B2SdzIzjzwkiWCCHOhTTOTy3-H2rNbbfs8jojm8AeaNYEfQCplT5aYjwtA',
  client_data_json := 'eyJ0eXBlIjoid2ViYXV0aG4uY3JlYXRlIiwiY2hhbGxlbmdlIjoiTU5XRVBjTkhfbjJUS0dkZWxrdC1fZTBmTVJLN0Q1S09iN0ZIX0liRlpMbyIsIm9yaWdpbiI6Imh0dHA6Ly9sb2NhbGhvc3QiLCJjcm9zc09yaWdpbiI6ZmFsc2UsIm90aGVyX2tleXNfY2FuX2JlX2FkZGVkX2hlcmUiOiJkbyBub3QgY29tcGFyZSBjbGllbnREYXRhSlNPTiBhZ2FpbnN0IGEgdGVtcGxhdGUuIFNlZSBodHRwczovL2dvby5nbC95YWJQZXgifQ',
  credential_at := '2020-12-15 08:09:00.20485+01'
);
ROLLBACK TO make_credential;

SELECT jsonb_pretty(webauthn.get_credentials(
  challenge := '\x014fc58eef9713a5c89e6094d5847faf39dd05efac8713a6024c5812e9178599'::bytea,
  user_name := 'alex.p.mueller@example.com',
  user_verification := 'discouraged',
  timeout := '00:05:00',
  relying_party_id := NULL,
  challenge_at := '2020-12-15 08:09:01.608713+01'
));

SAVEPOINT get_credentials;
SELECT jsonb_pretty(webauthn.get_credentials(
  challenge := '\x014fc58eef9713a5c89e6094d5847faf39dd05efac8713a6024c5812e9178599'::bytea,
  user_name := 'alex.p.mueller@example.com',
  user_verification := 'discouraged',
  timeout := '00:05:00',
  relying_party_id := NULL,
  challenge_at := '2020-12-15 08:09:01.608713+01'
));
ROLLBACK TO get_credentials;

SELECT * FROM webauthn.verify_assertion(
  credential_id := 'AXNBRMEOFaYGaROrEph1sOZ4kftILi9ry8vCw2fPQf712glIpQDRX-7HBQ2VmQVpRWU3A6Cu_XcKbnoC2SSy5_o0Z2qO7Owdnms8K0GsiqvWx3WtUPn0a8Ga6QWbkEvsUXOp9ikZ9v4DeYeTzzp0h2uAlx8ezayuqjB_uMQyB5kBVwRhkhZEmzQCl097',
  credential_type := 'public-key',
  authenticator_data := 'SZYN5YgOjGh0NBcPZHZgW4_krrmihjLHmVzzuoMdl2MFX9hhDw',
  client_data_json := 'eyJ0eXBlIjoid2ViYXV0aG4uZ2V0IiwiY2hhbGxlbmdlIjoiQVVfRmp1LVhFNlhJbm1DVTFZUl9yem5kQmUtc2h4T21Ba3hZRXVrWGhaayIsIm9yaWdpbiI6Imh0dHA6Ly9sb2NhbGhvc3QiLCJjcm9zc09yaWdpbiI6ZmFsc2V9',
  signature := 'MEYCIQCpphN2t04K-4uEft7Cm_xxcH1WuZuz3cSkGkMM7_bV-gIhAKwNJ9HYgZYpBJG-1xD6k9MRQU0J39HiqD7X_g9Wx_nU',
  user_handle := 'kn-QwjI3SNHiwkw5r30kBCP10dD77R13zP43Ky1nqyRPmIv1z3xy-Lfw_cs1lJipNcJvS4_B4rVktUGBqDRo2g',
  verified_at := '2020-12-15 08:09:03.3849+01'
);

SAVEPOINT verify_assertion;
SELECT * FROM webauthn.verify_assertion(
  credential_id := 'AXNBRMEOFaYGaROrEph1sOZ4kftILi9ry8vCw2fPQf712glIpQDRX-7HBQ2VmQVpRWU3A6Cu_XcKbnoC2SSy5_o0Z2qO7Owdnms8K0GsiqvWx3WtUPn0a8Ga6QWbkEvsUXOp9ikZ9v4DeYeTzzp0h2uAlx8ezayuqjB_uMQyB5kBVwRhkhZEmzQCl097',
  credential_type := 'public-key',
  authenticator_data := 'SZYN5YgOjGh0NBcPZHZgW4_krrmihjLHmVzzuoMdl2MFX9hhDw',
  client_data_json := 'eyJ0eXBlIjoid2ViYXV0aG4uZ2V0IiwiY2hhbGxlbmdlIjoiQVVfRmp1LVhFNlhJbm1DVTFZUl9yem5kQmUtc2h4T21Ba3hZRXVrWGhaayIsIm9yaWdpbiI6Imh0dHA6Ly9sb2NhbGhvc3QiLCJjcm9zc09yaWdpbiI6ZmFsc2V9',
  signature := 'MEYCIQCpphN2t04K-4uEft7Cm_xxcH1WuZuz3cSkGkMM7_bV-gIhAKwNJ9HYgZYpBJG-1xD6k9MRQU0J39HiqD7X_g9Wx_nU',
  user_handle := 'kn-QwjI3SNHiwkw5r30kBCP10dD77R13zP43Ky1nqyRPmIv1z3xy-Lfw_cs1lJipNcJvS4_B4rVktUGBqDRo2g',
  verified_at := '2020-12-15 08:09:03.3849+01'
);
ROLLBACK TO verify_assertion;

ROLLBACK;
