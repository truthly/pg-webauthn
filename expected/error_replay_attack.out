--
-- In this test we will simulate different "replay attacks",
-- by reusing values that are guaranteed to be unique
-- thanks to primary keys and unique constraints on columns.
--
-- We will use the SAVEPOINT ... ROLLBACK TO PostgreSQL feature,
-- to let errors happen that we want to test for,
-- but to roll back to the state before the error,
-- allowing the rest of the test to continue
-- which depends on the data produced so far.
-- Otherwise we would have to test all steps from start
-- in separate test files, which would be unnecessarily tedious.
--
-- See: https://www.postgresql.org/docs/current/sql-rollback-to.html
--
BEGIN;
CREATE EXTENSION IF NOT EXISTS webauthn CASCADE;
NOTICE:  installing required extension "pguecc"
NOTICE:  installing required extension "pgcrypto"
NOTICE:  installing required extension "cbor"
SELECT jsonb_pretty(webauthn.init_credential(
  challenge := '\x30d5843dc347fe7d9328675e964b7efded1f3112bb0f928e6fb147fc86c564ba'::bytea,
  user_name := 'alex.p.mueller@example.com',
  user_id := '\x927f90c2323748d1e2c24c39af7d240423f5d1d0fbed1d77ccfe372b2d67ab244f988bf5cf7c72f8b7f0fdcb359498a935c26f4b8fc1e2b564b54181a83468da'::bytea,
  user_display_name := 'Alex P. Müller',
  relying_party_name := 'ACME Corporation',
  relying_party_id := NULL,
  user_verification := 'discouraged',
  timeout := '00:05:00'::interval,
  challenge_at := '2020-12-15 08:08:56.596311+01'
));
                                                jsonb_pretty                                                 
-------------------------------------------------------------------------------------------------------------
 {                                                                                                          +
     "publicKey": {                                                                                         +
         "rp": {                                                                                            +
             "name": "ACME Corporation"                                                                     +
         },                                                                                                 +
         "user": {                                                                                          +
             "id": "kn-QwjI3SNHiwkw5r30kBCP10dD77R13zP43Ky1nqyRPmIv1z3xy-Lfw_cs1lJipNcJvS4_B4rVktUGBqDRo2g",+
             "name": "alex.p.mueller@example.com",                                                          +
             "displayName": "Alex P. Müller"                                                                +
         },                                                                                                 +
         "timeout": 300000,                                                                                 +
         "challenge": "MNWEPcNH_n2TKGdelkt-_e0fMRK7D5KOb7FH_IbFZLo",                                        +
         "attestation": "none",                                                                             +
         "pubKeyCredParams": [                                                                              +
             {                                                                                              +
                 "alg": -7,                                                                                 +
                 "type": "public-key"                                                                       +
             }                                                                                              +
         ],                                                                                                 +
         "authenticatorSelection": {                                                                        +
             "userVerification": "discouraged",                                                             +
             "requireResidentKey": false                                                                    +
         }                                                                                                  +
     }                                                                                                      +
 }
(1 row)

SAVEPOINT init_credential;
SELECT jsonb_pretty(webauthn.init_credential(
  challenge := '\x30d5843dc347fe7d9328675e964b7efded1f3112bb0f928e6fb147fc86c564ba'::bytea,
  user_name := 'alex.p.mueller@example.com',
  user_id := '\x927f90c2323748d1e2c24c39af7d240423f5d1d0fbed1d77ccfe372b2d67ab244f988bf5cf7c72f8b7f0fdcb359498a935c26f4b8fc1e2b564b54181a83468da'::bytea,
  user_display_name := 'Alex P. Müller',
  relying_party_name := 'ACME Corporation',
  relying_party_id := NULL,
  user_verification := 'discouraged',
  timeout := '00:05:00'::interval,
  challenge_at := '2020-12-15 08:08:56.596311+01'
));
ERROR:  duplicate key value violates unique constraint "credential_challenges_pkey"
DETAIL:  Key (challenge)=(\x30d5843dc347fe7d9328675e964b7efded1f3112bb0f928e6fb147fc86c564ba) already exists.
CONTEXT:  SQL function "init_credential" statement 1
ROLLBACK TO init_credential;
SELECT * FROM webauthn.make_credential(
  credential_id := 'AXNBRMEOFaYGaROrEph1sOZ4kftILi9ry8vCw2fPQf712glIpQDRX-7HBQ2VmQVpRWU3A6Cu_XcKbnoC2SSy5_o0Z2qO7Owdnms8K0GsiqvWx3WtUPn0a8Ga6QWbkEvsUXOp9ikZ9v4DeYeTzzp0h2uAlx8ezayuqjB_uMQyB5kBVwRhkhZEmzQCl097',
  credential_type := 'public-key',
  attestation_object := 'o2NmbXRkbm9uZWdhdHRTdG10oGhhdXRoRGF0YVkBEUmWDeWIDoxodDQXD2R2YFuP5K65ooYyx5lc87qDHZdjRV_YYQytzgACNbzGCmSLCyXx8FUDAI0Bc0FEwQ4VpgZpE6sSmHWw5niR-0guL2vLy8LDZ89B_vXaCUilANFf7scFDZWZBWlFZTcDoK79dwpuegLZJLLn-jRnao7s7B2eazwrQayKq9bHda1Q-fRrwZrpBZuQS-xRc6n2KRn2_gN5h5PPOnSHa4CXHx7NrK6qMH-4xDIHmQFXBGGSFkSbNAKXT3ulAQIDJiABIVggxMJIqQI-xxUSsJGR9HIbIZHKIiN-alA9B2SdzIzjzwkiWCCHOhTTOTy3-H2rNbbfs8jojm8AeaNYEfQCplT5aYjwtA',
  client_data_json := 'eyJ0eXBlIjoid2ViYXV0aG4uY3JlYXRlIiwiY2hhbGxlbmdlIjoiTU5XRVBjTkhfbjJUS0dkZWxrdC1fZTBmTVJLN0Q1S09iN0ZIX0liRlpMbyIsIm9yaWdpbiI6Imh0dHA6Ly9sb2NhbGhvc3QiLCJjcm9zc09yaWdpbiI6ZmFsc2UsIm90aGVyX2tleXNfY2FuX2JlX2FkZGVkX2hlcmUiOiJkbyBub3QgY29tcGFyZSBjbGllbnREYXRhSlNPTiBhZ2FpbnN0IGEgdGVtcGxhdGUuIFNlZSBodHRwczovL2dvby5nbC95YWJQZXgifQ',
  credential_at := '2020-12-15 08:09:00.20485+01'
);
                                                              user_id                                                               
------------------------------------------------------------------------------------------------------------------------------------
 \x927f90c2323748d1e2c24c39af7d240423f5d1d0fbed1d77ccfe372b2d67ab244f988bf5cf7c72f8b7f0fdcb359498a935c26f4b8fc1e2b564b54181a83468da
(1 row)

SAVEPOINT make_credential;
SELECT * FROM webauthn.make_credential(
  credential_id := 'AXNBRMEOFaYGaROrEph1sOZ4kftILi9ry8vCw2fPQf712glIpQDRX-7HBQ2VmQVpRWU3A6Cu_XcKbnoC2SSy5_o0Z2qO7Owdnms8K0GsiqvWx3WtUPn0a8Ga6QWbkEvsUXOp9ikZ9v4DeYeTzzp0h2uAlx8ezayuqjB_uMQyB5kBVwRhkhZEmzQCl097',
  credential_type := 'public-key',
  attestation_object := 'o2NmbXRkbm9uZWdhdHRTdG10oGhhdXRoRGF0YVkBEUmWDeWIDoxodDQXD2R2YFuP5K65ooYyx5lc87qDHZdjRV_YYQytzgACNbzGCmSLCyXx8FUDAI0Bc0FEwQ4VpgZpE6sSmHWw5niR-0guL2vLy8LDZ89B_vXaCUilANFf7scFDZWZBWlFZTcDoK79dwpuegLZJLLn-jRnao7s7B2eazwrQayKq9bHda1Q-fRrwZrpBZuQS-xRc6n2KRn2_gN5h5PPOnSHa4CXHx7NrK6qMH-4xDIHmQFXBGGSFkSbNAKXT3ulAQIDJiABIVggxMJIqQI-xxUSsJGR9HIbIZHKIiN-alA9B2SdzIzjzwkiWCCHOhTTOTy3-H2rNbbfs8jojm8AeaNYEfQCplT5aYjwtA',
  client_data_json := 'eyJ0eXBlIjoid2ViYXV0aG4uY3JlYXRlIiwiY2hhbGxlbmdlIjoiTU5XRVBjTkhfbjJUS0dkZWxrdC1fZTBmTVJLN0Q1S09iN0ZIX0liRlpMbyIsIm9yaWdpbiI6Imh0dHA6Ly9sb2NhbGhvc3QiLCJjcm9zc09yaWdpbiI6ZmFsc2UsIm90aGVyX2tleXNfY2FuX2JlX2FkZGVkX2hlcmUiOiJkbyBub3QgY29tcGFyZSBjbGllbnREYXRhSlNPTiBhZ2FpbnN0IGEgdGVtcGxhdGUuIFNlZSBodHRwczovL2dvby5nbC95YWJQZXgifQ',
  credential_at := '2020-12-15 08:09:00.20485+01'
);
ERROR:  duplicate key value violates unique constraint "credentials_pkey"
DETAIL:  Key (credential_id)=(\x01734144c10e15a6066913ab129875b0e67891fb482e2f6bcbcbc2c367cf41fef5da0948a500d15feec7050d9599056945653703a0aefd770a6e7a02d924b2e7fa34676a8eecec1d9e6b3c2b41ac8aabd6c775ad50f9f46bc19ae9059b904bec5173a9f62919f6fe03798793cf3a74876b80971f1ecdacaeaa307fb8c4320799015704619216449b3402974f7b) already exists.
CONTEXT:  SQL function "make_credential" statement 1
ROLLBACK TO make_credential;
SELECT jsonb_pretty(webauthn.get_credentials(
  challenge := '\x014fc58eef9713a5c89e6094d5847faf39dd05efac8713a6024c5812e9178599'::bytea,
  user_name := 'alex.p.mueller@example.com',
  user_verification := 'discouraged',
  timeout := '00:05:00',
  relying_party_id := NULL,
  challenge_at := '2020-12-15 08:09:01.608713+01'
));
                                                                                                     jsonb_pretty                                                                                                      
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {                                                                                                                                                                                                                    +
     "publicKey": {                                                                                                                                                                                                   +
         "timeout": 300000,                                                                                                                                                                                           +
         "challenge": "AU_Fju-XE6XInmCU1YR_rzndBe-shxOmAkxYEukXhZk",                                                                                                                                                  +
         "allowCredentials": [                                                                                                                                                                                        +
             {                                                                                                                                                                                                        +
                 "id": "AXNBRMEOFaYGaROrEph1sOZ4kftILi9ry8vCw2fPQf712glIpQDRX-7HBQ2VmQVpRWU3A6Cu_XcKbnoC2SSy5_o0Z2qO7Owdnms8K0GsiqvWx3WtUPn0a8Ga6QWbkEvsUXOp9ikZ9v4DeYeTzzp0h2uAlx8ezayuqjB_uMQyB5kBVwRhkhZEmzQCl097",+
                 "type": "public-key"                                                                                                                                                                                 +
             }                                                                                                                                                                                                        +
         ],                                                                                                                                                                                                           +
         "userVerification": "discouraged"                                                                                                                                                                            +
     }                                                                                                                                                                                                                +
 }
(1 row)

SAVEPOINT get_credentials;
SELECT jsonb_pretty(webauthn.get_credentials(
  challenge := '\x014fc58eef9713a5c89e6094d5847faf39dd05efac8713a6024c5812e9178599'::bytea,
  user_name := 'alex.p.mueller@example.com',
  user_verification := 'discouraged',
  timeout := '00:05:00',
  relying_party_id := NULL,
  challenge_at := '2020-12-15 08:09:01.608713+01'
));
ERROR:  duplicate key value violates unique constraint "assertion_challenges_pkey"
DETAIL:  Key (challenge)=(\x014fc58eef9713a5c89e6094d5847faf39dd05efac8713a6024c5812e9178599) already exists.
CONTEXT:  SQL function "get_credentials" statement 1
ROLLBACK TO get_credentials;
SELECT * FROM webauthn.verify_assertion(
  credential_id := 'AXNBRMEOFaYGaROrEph1sOZ4kftILi9ry8vCw2fPQf712glIpQDRX-7HBQ2VmQVpRWU3A6Cu_XcKbnoC2SSy5_o0Z2qO7Owdnms8K0GsiqvWx3WtUPn0a8Ga6QWbkEvsUXOp9ikZ9v4DeYeTzzp0h2uAlx8ezayuqjB_uMQyB5kBVwRhkhZEmzQCl097',
  credential_type := 'public-key',
  authenticator_data := 'SZYN5YgOjGh0NBcPZHZgW4_krrmihjLHmVzzuoMdl2MFX9hhDw',
  client_data_json := 'eyJ0eXBlIjoid2ViYXV0aG4uZ2V0IiwiY2hhbGxlbmdlIjoiQVVfRmp1LVhFNlhJbm1DVTFZUl9yem5kQmUtc2h4T21Ba3hZRXVrWGhaayIsIm9yaWdpbiI6Imh0dHA6Ly9sb2NhbGhvc3QiLCJjcm9zc09yaWdpbiI6ZmFsc2V9',
  signature := 'MEYCIQCpphN2t04K-4uEft7Cm_xxcH1WuZuz3cSkGkMM7_bV-gIhAKwNJ9HYgZYpBJG-1xD6k9MRQU0J39HiqD7X_g9Wx_nU',
  user_handle := 'kn-QwjI3SNHiwkw5r30kBCP10dD77R13zP43Ky1nqyRPmIv1z3xy-Lfw_cs1lJipNcJvS4_B4rVktUGBqDRo2g',
  verified_at := '2020-12-15 08:09:03.3849+01'
);
                                                              user_id                                                               
------------------------------------------------------------------------------------------------------------------------------------
 \x927f90c2323748d1e2c24c39af7d240423f5d1d0fbed1d77ccfe372b2d67ab244f988bf5cf7c72f8b7f0fdcb359498a935c26f4b8fc1e2b564b54181a83468da
(1 row)

SAVEPOINT verify_assertion;
SELECT * FROM webauthn.verify_assertion(
  credential_id := 'AXNBRMEOFaYGaROrEph1sOZ4kftILi9ry8vCw2fPQf712glIpQDRX-7HBQ2VmQVpRWU3A6Cu_XcKbnoC2SSy5_o0Z2qO7Owdnms8K0GsiqvWx3WtUPn0a8Ga6QWbkEvsUXOp9ikZ9v4DeYeTzzp0h2uAlx8ezayuqjB_uMQyB5kBVwRhkhZEmzQCl097',
  credential_type := 'public-key',
  authenticator_data := 'SZYN5YgOjGh0NBcPZHZgW4_krrmihjLHmVzzuoMdl2MFX9hhDw',
  client_data_json := 'eyJ0eXBlIjoid2ViYXV0aG4uZ2V0IiwiY2hhbGxlbmdlIjoiQVVfRmp1LVhFNlhJbm1DVTFZUl9yem5kQmUtc2h4T21Ba3hZRXVrWGhaayIsIm9yaWdpbiI6Imh0dHA6Ly9sb2NhbGhvc3QiLCJjcm9zc09yaWdpbiI6ZmFsc2V9',
  signature := 'MEYCIQCpphN2t04K-4uEft7Cm_xxcH1WuZuz3cSkGkMM7_bV-gIhAKwNJ9HYgZYpBJG-1xD6k9MRQU0J39HiqD7X_g9Wx_nU',
  user_handle := 'kn-QwjI3SNHiwkw5r30kBCP10dD77R13zP43Ky1nqyRPmIv1z3xy-Lfw_cs1lJipNcJvS4_B4rVktUGBqDRo2g',
  verified_at := '2020-12-15 08:09:03.3849+01'
);
ERROR:  duplicate key value violates unique constraint "assertions_pkey"
DETAIL:  Key (signature)=(\x3046022100a9a61376b74e0afb8b847edec29bfc71707d56b99bb3ddc4a41a430ceff6d5fa022100ac0d27d1d88196290491bed710fa93d311414d09dfd1e2a83ed7fe0f56c7f9d4) already exists.
CONTEXT:  SQL function "verify_assertion" statement 1
ROLLBACK TO verify_assertion;
ROLLBACK;
